confluentinc/ksql
# feat: support JSON schema inference in ksqlDB

### Description 

This is the first cut toward support JSON schemas in ksqlDB. Creating a topic with `C*` and `JSON` no longer requires specifying the schema if the schema exists in schema registry, but when you create a JSON topic (either with `C*AS` or `C*`) the schema is not registered into schema-registry.

~This PR is pretty straightforward, but there's one thing to understand - the schema registry converters add a magic byte (`0x00`) and four bytes to represent the schema ID at the beginning of every message. I made the choice to rely on standard JSON encoding - so we serialize without the magic byte/schemaID and when we deserialize we make sure to cut off the first 5 bytes from non-null events.~

~The only concern with this approach (cc @derekjn @MichaelDrogalis @colinhicks) is that even if data coming into ksqlDB is serialized with the `JsonSchemaConverter` (schema registry's format), data coming out will be vanilla JSON - this means that the `JsonSchemaConverter` will throw a `missing magic byte` error when trying to deserialize data coming out of ksqlDB.~

To allow users to output the data in the schema registry format, we will later introduce a new Format: `JSON_SR` which will (like `JSON`) be able to read either format, but will output data with the magic byte and schemaID prepended.

### Testing done 

1. Create a topic with JSON schema registered
```sh
> ./bin/kafka-json-schema-console-producer --broker-list localhost:9092 --topic json --property value.schema='{"type":"object","properties":{"f1":{"type":"string"}}}'
{"f1":"foo"}
```
2. Import that stream to ksqlDB (see #4556):
```
ksql> CREATE STREAM json WITH(KAFKA_TOPIC='json', VALUE_FORMAT='json');

 Message
----------------
 Stream created
----------------
ksql> DESCRIBE EXTENDED JSON;

Name                 : JSON
Type                 : STREAM
Key field            :
Timestamp field      : Not set - using <ROWTIME>
Key format           : KAFKA
Value format         : JSON
Kafka topic          : json (partitions: 1, replication: 1)
Statement            : CREATE STREAM JSON (F1 STRING) WITH (AVRO_SCHEMA_ID=2, KAFKA_TOPIC='json', VALUE_FORMAT='json');

 Field   | Type
-------------------------------------
 ROWTIME | BIGINT           (system)
 ROWKEY  | VARCHAR(STRING)  (system)
 F1      | VARCHAR(STRING)
-------------------------------------
```
3. `INSERT INTO` (note that this uses the standard JSON serialization, so we will have a topic with _mixed_ serializations) also see (#4555):
```
ksql> INSERT INTO json (f1) VALUES ('bar');
ksql> PRINT json FROM BEGINNING;
Format:STRING
2/13/20 11:02:26 AM PST , NULL , \x00\x00\x00\x00\x02{"f1":"foo"}
2/13/20 11:36:55 AM PST , NULL , {"F1":"bar"}
```
4. Make sure we can read from it:
```
ksql> SELECT * FROM JSON EMIT CHANGES;
+------------------------------+------------------------------+------------------------------+
|ROWTIME                       |ROWKEY                        |F1                            |
+------------------------------+------------------------------+------------------------------+
|1581620546467                 |null                          |foo                           |
|1581622615960                 |null                          |bar                           |
```

### Reviewer checklist
